remove solution tag from inputtypes and update default templates

muhammad-ammar opened the pull request:
##### [TNL-4822](https://openedx.atlassian.net/browse/TNL-4822) [FEDX-208](https://openedx.atlassian.net/browse/FEDX-208)

#### [Sandbox](https://studio-capatemplates.sandbox.edx.org)

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @muzaffaryousaf   
- [x] Code review: @ehteshamkafeel  
- [x] Code review: @cahrens   


### Post-review
- [x] Squash commits

### Changes

- Update the default YAML templates as per the new format
- Move the `<solution>` tag inside responsetype

muhammad-ammar commented:
@muzaffaryousaf @ehteshamkafeel This is ready for review.

muzaffaryousaf commented:
@muhammad-ammar Do we have sandbox for this ?

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/spec/problem/edit_spec.coffee:
```diff
@@ -271,29 +270,27 @@ describe 'MarkdownEditingDescriptor', ->
+                <solution>
+                    <div class="detailed-solution">
+                    <p>Explanation</p>
+                    <p>The release of the iPod allowed consumers to carry their entire music library with them in a format that did not rely on fragile and energy-intensive spinning disks.</p>
```
Indentation ?

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/spec/problem/edit_spec.coffee:
```diff
@@ -231,28 +231,27 @@ describe 'MarkdownEditingDescriptor', ->
+            <multiplechoiceresponse>
+                <p>A multiple choice problem presents radio buttons for student input. Students can only select a single option presented. Multiple Choice questions have been the subject of many areas of research due to the early invention and adoption of bubble sheets.</p>
+                <p>One of the main elements that goes into a good multiple choice question is the existence of good distractors. That is, each of the alternate responses presented to the student should be the result of a plausible mistake that a student might make.</p>
+                <p>What Apple device competed with the portable CD player?</p>
```
This is not being turned into label because we we don't have `>><<` formatters around question title ?

muzaffaryousaf replied:
I would prefer that we add some tests with `>><<` if they're not present already.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/spec/problem/edit_spec.coffee:
```diff
@@ -363,22 +360,20 @@ describe 'MarkdownEditingDescriptor', ->
+                    <p>Explanation</p>
+                    <p>Multiple Choice also allows students to select from a variety of pre-written responses, although the format makes it easier for students to read very long response options. Optionresponse also differs slightly because students are more likely to think of an answer and then search for it rather than relying purely on recognition to answer the question.</p>
+                    </div>
+                </solution>
```
NIT: Can you add indentation to all the solution tag's content. It's almost in every test in this file.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/src/problem/edit.coffee:
```diff
@@ -583,7 +583,7 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
 
       // these will be placed at outside the end of responsetype
-      var independentTagNames = ['solution'];
+      var independentTagNames = [];
```
Do we need this functionality at all ?

cahrens replied:
I agree-- let's get rid of independentTagNames and independentTagNodes if they are not needed (they do not seem to be in master).

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/circuitschematic.yaml:
```diff
@@ -60,24 +73,17 @@ data: |
+                    <p><img src="/static/images/high_pass_filter.png" alt=""/></p>
+                </div>
+            </solution>
+        </schematicresponse>
```
NVM. It does't have any Markdown. :)

muzaffaryousaf replied:
Do we only need to move `label` inside the responseType ?

muhammad-ammar replied:
@muzaffaryousaf I have moved only those tags inside which are most relevant to responsetype. The `<p>` tags in `circuitschematic.yaml` are problem wide description.

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/customgrader.yaml:
```diff
@@ -4,32 +4,34 @@ metadata:
+            <p>You can use the following example problem as a model.</p>
+            <hr/>
 
       <script type="loncapa/python">
```
NIT: Can you indent this accordingly.

muzaffaryousaf replied:
What's up with this ? Don't we need to put the things inside responseTypes ?

muzaffaryousaf commented:
@muhammad-ammar Can you explain what kind of changes you made to the Non-CAPA problems. I can see that you converted the question title from `p` to `label` and moving the solution tag inside the response tag but how about moving the rest of the content inside the responseTypes ?

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/string_response_hint.yaml:
```diff
@@ -24,31 +24,22 @@ metadata:
+                <p>Be sure to select Settings to specify a Display Name and other values that apply.</p>
+                <p>Use the following example problem as a model.</p>
+                <label>Which U.S. state has the largest land area?</label>
+                <correcthint>Alaska is 576,400 square miles, more than double the land area of the second largest state, Texas.</correcthint>
```
Indentation.

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/problem_with_hint_in_latex.yaml:
```diff
@@ -81,12 +81,10 @@ data: |
+        <label>What is the best programming language that exists today? You may enter your answer in upper or lower case, with or without quotes.</label>
+          <textline correct_answer="python"/>
+          <hintgroup hintfn="hint_fn"/>
+        </customresponse>
```
Indentation.

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/latex_problem.yaml:
```diff
@@ -88,37 +88,41 @@ metadata:
-      <p>For more information, see
-      <a href="http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/exercises_tools/problem_in_latex.html" target="_blank">
-      Problem Written in LaTeX</a> in <i>Building and Running an edX Course</i>.</p>
+    <problem>
```
Why we removed the attributes from problem tag ?

muhammad-ammar replied:
@muzaffaryousaf They are not used by CAPA, they are used during import/export. Someone incorrect set these here.

muzaffaryousaf replied:
For safe side, have you tried you import/export course without these attributes ?

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/formularesponse.yaml:
```diff
@@ -3,54 +3,58 @@ metadata:
+            <responseparam type="tolerance" default="0.00001"/>
+            <formulaequationinput size="40"  label="Enter the equation"/>
+          </formularesponse>
+      </problem>
```
Need to fix indentation in this file.

muzaffaryousaf commented:
@muhammad-ammar I'm done with first pass, generally it looks good. Ping me back when you're done with addressing my questions/feedback.

muzaffaryousaf commented:
@ehteshamkafeel Can you review this first ?

muzaffaryousaf commented:
@cahrens Would you mind to give it a quick look ?
@sstack22 @lamagnifica @cptvitamin @chris-mike  FYI ^^

cahrens commented on lms/templates/problem.html:
```diff
@@ -12,9 +12,7 @@ <h3 class="hd hd-2 problem-header">
-    <div aria-live="polite">
-        ${ HTML(problem['html']) }
-    </div>
+  ${ HTML(problem['html']) }
```
This is OK'd by @cptvitamin?

muhammad-ammar replied:
@cahrens yes

cahrens commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response.yaml:
```diff
@@ -21,27 +21,26 @@ metadata:
+                <p>When you add the problem, be sure to select Settings to specify a Display Name and other values that apply.</p>
+                <p>You can use the following example problem as a model.</p>
+                <label>The following languages are in the Indo-European family:</label>
+                <checkboxgroup label="The following languages are in the Indo-European family:">
```
Have you talked to @cptvitamin about avoiding this type of duplication?

muhammad-ammar replied:
This will be handled when we work on FEDX-200.

muzaffaryousaf replied:
I've added this to FEDX-200, we can cater this along with question description.

cahrens commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response.yaml:
```diff
@@ -21,27 +21,26 @@ metadata:
+                    <choice correct="true">French</choice>
+                    <choice correct="false">Hungarian</choice>
+                </checkboxgroup>
+                <p>Note: Make sure you select all of the correct options—there may be more than one!</p>
```
This note is a bit odd-- I think it won't be accessible. Should it be to descriptive element on the label?

muzaffaryousaf replied:
@cahrens We're going to move this text in question description with "Note:" in upcoming descriptions PR.

lamagnifica replied:
+1

cahrens commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response_hint.yaml:
```diff
@@ -28,42 +28,39 @@ metadata:
+                <p>Be sure to select Settings to specify a Display Name and other values that apply.</p>
+                <p>Use the following example problem as a model.</p>
+                <label>Which of the following is a fruit? Check all that apply.</label>
+                <checkboxgroup label="Which of the following is a fruit? Check all that apply.">
```
Again, question about the duplication.

cahrens commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response.yaml:
```diff
@@ -21,27 +21,26 @@ metadata:
+        <problem>
+            <choiceresponse>
+                <p>Checkbox problems allow learners to select multiple options. Learners can see all the options along with the problem text.</p>
+                <p>When you add the problem, be sure to select Settings to specify a Display Name and other values that apply.</p>
```
Why did this lose the strong tags around Display Name and Settings? Does it match what is specified by the markdown?

lamagnifica replied:
When you add the problem ^component, be sure...  (? the Settings > Display Name instruction applies to the entire component, was this changed for a reason?)
also, can the bolding be added back, or was that deliberately removed? ("Settings" and "Display Name" were in  bold font and have initial capital letters to draw attention, since they are labels in the UI.)

cahrens commented on common/lib/xmodule/xmodule/templates/problem/multiplechoice.yaml:
```diff
@@ -22,32 +22,30 @@ metadata:
+        <problem>
+           <multiplechoiceresponse>
+               <p>Multiple choice problems allow learners to select only one option. Learners can see all the options along with the problem text.</p>
+               <p>When you add the problem, be sure to select Settings to specify a Display Name and other values that apply.</p>
```
Here is another case where the strong tags got lost. Again, my main concern is that the OLX matches what would be generated from the markdown.

cahrens commented on common/lib/xmodule/xmodule/templates/problem/multiplechoice.yaml:
```diff
@@ -22,32 +22,30 @@ metadata:
+               <p>When you add the problem, be sure to select Settings to specify a Display Name and other values that apply.</p>
+               <p>You can use the following example problem as a model.</p>
+               <label>Which of the following countries has the largest population?</label>
+               <choicegroup label="Which of the following countries has the largest population?" type="MultipleChoice">
```
And same question about the duplication.

cahrens commented:
I'm done reviewing-- looking good!

muhammad-ammar commented:
> Can you explain what kind of changes you made to the Non-CAPA problems. I can see that you converted the question title from p to label and moving the solution tag inside the response tag but how about moving the rest of the content inside the responseTypes ?

@muzaffaryousaf For Non-CAPA, i have only moved those things inside the responsetype which are question specific. In Non-CAPA problems we have code snippets used for grading, i am not sure if we need to move them inside responsetype, for some problems they are outside and for others they are inside. BTW i have moved code snippets inside the responsetype but I am not sure what is the correct behavior?

muhammad-ammar commented:
jenkins run all

ehteshamkafeel commented:
Looks Good to me 👍

muhammad-ammar commented:
@muzaffaryousaf @cahrens I have addressed your feedback. You can continue the review.

Here is the [Sandbox](https://studio-capatemplates.sandbox.edx.org)

muhammad-ammar commented:
jenkins run bokchoy

cahrens commented on common/lib/xmodule/xmodule/templates/problem/numericalresponse.yaml:
```diff
@@ -15,44 +15,46 @@ metadata:
+            <numericalresponse answer="9.3*10^7">
+                <p>In a numerical input problem, learners enter numbers or a specific and relatively simple mathematical expression. Learners enter the response in plain text, and the system then converts the text to a symbolic expression that learners can see below the response field.</p>
+                <p>The system can handle several types of characters, including basic operators, fractions, exponents, and common constants such as i. You can refer learners to
+                    <a href="http://edx.readthedocs.io/projects/edx-guide-for-students/en/latest/completing_assignments/SFD_mathformatting.html#math-formatting" target="_blank">Entering Mathematical and Scientific Expressions</a>in the<i> EdX Learner's Guide</i>
```
Having an href in a markdown problem doesn't make sense to me. There is no way to specify it markdown, so if you bring up the Advanced Editor and then save it, the link goes away (as it is only specified here in the OLX).

@alisan617 What do you think of removing the hyperlink? I'd like what is displayed to match what is in the markdown. Note though that assuming this is a pre-existing situation, it doesn't have to be resolved for this PR.

muzaffaryousaf replied:
This is something we can do for all problem types. i.e  to make mark down and OLX in sync with each other. I'll create a separate ticket for this and i think it's not a high priority for now.

cahrens replied:
Again, sorry, @alisan617 --- I meant to tag @lamagnifica instead! My Alison/Alisan confusion is beginning....

lamagnifica replied:
FWIW, the link works fine when you add a new unit and then add a Numerical Input problem to it. I understand your concern about what happens when you switch to the advanced editor and save, but the title of the specific referenced topic does continue to appear, although no longer hyperlinked, and it might still be useful. 
So sure, we can take the hrefs out of all of these yaml files and add them back at another time, but I don't think they're doing any harm and might provide some useful references to additional information.

cahrens commented on common/lib/xmodule/xmodule/templates/problem/string_response.yaml:
```diff
@@ -18,26 +18,21 @@ metadata:
+        <problem>
+            <stringresponse answer="Nanjing Higher Normal Institute" type="ci">
+                <p>In text input problems, also known as "fill-in-the-blank" problems, learners enter text into a response field. The text that the learner enters must match your specified answer text exactly. You can specify more than one correct answer. Learners must enter a response that matches one of the correct answers exactly.</p>
+                <p>When you add the problem, be sure to select <strong>Settings</strong> to specify a <strong>Display Name</strong> and other values that apply.</p>
```
@alisan617 This is another place where the display OLX does not match the markdown-- "Settings" and "Display Name" are not marked as bold in the markdown (I don't think there is a syntax for that), and therefore if you edit the problem in any way and save it, the bold text goes away. I think this is very confusing, and the OLX should be changed to match the markdown.

alisan617 replied:
** can make markdown bold. I have no context what this is. Maybe put a quote around it, or make the text all caps, if this is for visual purpose only.

cahrens replied:
If ** can make text bold, please make the markdown match the OLX, @muhammad-ammar.

Sorry, @alisan617 --- I meant to tag @lamagnifica instead! My Alison/Alisan confusion is beginning....

lamagnifica replied:
ah, ok, see my comments earlier in this PR about why Display Name and Settings were presented in bold and initial caps. 
Alisan's suggestion is a good one, but I just tested in Studio and unfortunately our lame markdown does not support the use of ** to bold a word.
Like the question about href support in markdown, my take is that the initial presentation is as desired, and while saving removes the styling, it doesn't hurt anything to leave it in. Maybe we could just leave it alone until our weak markdown can support bold (or italic, or underlining...)

cahrens replied:
@lamagnifica My concern is that authors see styling on these problems that they think they can achieve via markdown, but that isn't the case. I feel pretty strongly that the OLX should match the markdown.

lamagnifica replied:
Understood. Let me come up with an alternate wording/styling solution...

lamagnifica replied:
hello @muhammad-ammar here is my suggestion. The guidance about the templates on the wiki page (https://openedx.atlassian.net/wiki/pages/viewpage.action?pageId=30966057) is very helpful. Let's simplify these templates a lot!
Rather than try to draft them in PR comments, I will add text to the child wiki pages (like https://openedx.atlassian.net/wiki/display/A11Y/Dropdown) and invite comment there.

cahrens commented:
👍  Though I think some cleanup needs to be done to the problems with markdown to ensure that the markdown matches what is actually displayed (the initial value of the OLX). Perhaps for each of the "simple problems", just bring up the Advanced Editor for each and use that generated OLX to ensure that it matches the markdown.

muzaffaryousaf commented:
@cahrens @sstack22 I've created this ticket to make markdown and OLX in sync. https://openedx.atlassian.net/browse/FEDX-209

muzaffaryousaf commented:
@sstack22 Here is how the default templates will look like. You can see them on the sandbox, if you like.

muzaffaryousaf commented:
@muhammad-ammar On sandbox the indentation is bit off when we see the OLX in advance Editor, is this something we still need to address ?

muzaffaryousaf commented on common/lib/xmodule/xmodule/templates/problem/optionresponse.yaml:
```diff
@@ -15,22 +15,19 @@ metadata:
+                <p>Dropdown problems allow learners to select only one option from a list of options.</p>
+                <p>When you add the problem, be sure to select Settings to specify a Display Name and other values that apply.</p>
+                <p>You can use the following example problem as a model.</p>
+                <label>Which of the following countries celebrates its independence on August 15?</label>
```
Can we remove white space before `>>` formatter on this line https://github.com/edx/edx-platform/pull/12912/files#diff-99c8ef2c886a1b7d1635f7fe8e404402R11 This looks a bit off in UI. Also, we have to super careful about the white spaces just like this.

muhammad-ammar replied:
@muzaffaryousaf This space is already there in [master](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/templates/problem/optionresponse.yaml).

muzaffaryousaf replied:
Yeah i've seen that. Looks like the space is not intended so we can remove it.

muzaffaryousaf commented on common/lib/xmodule/xmodule/js/spec/problem/edit_spec.coffee:
```diff
@@ -766,52 +749,46 @@ describe 'MarkdownEditingDescriptor', ->
+                <label>Which of the following countries has the largest population?</label>
+                <choicegroup label="Which of the following countries has the largest population?" type="MultipleChoice">
+                    <choice correct="false">Brazil <choicehint>timely feedback -- explain why an almost correct answer is wrong</choicehint>
+                    </choice>
```
Can we move this to the end of the same line ?  This looks a bit weird in UI.

muhammad-ammar replied:
@muzaffaryousaf UI? This is a spec file.

muzaffaryousaf replied:
My bad. :)

muzaffaryousaf commented:
LGTM 👍 except few nits.

muzaffaryousaf commented:
@sstack22 Can you please confirm that we're good with default templates especially putting the `p` tags in their respective responseTypes for non-CAPA problems.

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/checkboxes_response_hint.yaml:
```diff
@@ -28,42 +28,39 @@ metadata:
+                <p>You can provide feedback for each option in a checkbox problem, with distinct feedback depending on whether or not the learner selects that option.</p>
+                <p>You can also provide compound feedback for a specific combination of answers. For example, if you have three possible answers in the problem, you can configure specific feedback for when a learner selects each combination of possible answers.</p>
+                <p>You can also add hints for learners.</p>
+                <p>Be sure to select Settings to specify a Display Name and other values that apply.</p>
```
same comment about adding the bold font back (throughout)

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/drag_and_drop.yaml:
```diff
@@ -5,21 +5,22 @@ metadata:
+      <hr/>
+      <customresponse>
+          <h3>Simple Drag and Drop</h3>
+          <p>Drag each word in the scrollbar to the bucket that matches the number of letters in the word.</p>
```
<label> doesn't apply to the deprecated drag and drop problems?

muzaffaryousaf replied:
@lamagnifica @muhammad-ammar I just came to know that we won't be making any changes to the Drag & Drop problems since it's deprecated and open craft is working on a new component for this. @sstack22 right?

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/imageresponse.yaml:
```diff
@@ -2,34 +2,29 @@
+                in
+                <i>Building and Running an edx Course</i>.
+            </p>
+            <p>When you add the problem, be sure to select
```
odd line feeding here (IMO)

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/latex_problem.yaml:
```diff
@@ -88,37 +88,41 @@ metadata:
-          <optionresponse>
-            <optioninput options="('India','Spain','China','Bermuda')" correct="India"></optioninput>
-          </optionresponse>
+      <p><strong>Example Option Problem</strong></p>
```
should this actually be a heading rather than a bolded paragraph?

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/latex_problem.yaml:
```diff
@@ -88,37 +88,41 @@ metadata:
+      </optionresponse>
+      <br/>
+
+      <p><strong>Example Multiple Choice Problem</strong></p>
```
same as line 103-104

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/latex_problem.yaml:
```diff
@@ -169,60 +196,55 @@ data: |
+      <br/>
+
+      <p><strong>Example Hidden Explanation</strong></p>
+      <p>You can provide additional information that only appears at certain times by including a "showhide" flag. </p>
```
this one really doesn't have a question, does it. hmm.

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/multiplechoice_hint.yaml:
```diff
@@ -21,26 +21,32 @@ metadata:
+            <multiplechoiceresponse>
+                <p>You can provide feedback for each option in a multiple choice problem.</p>
+                <p>You can also add hints for learners.</p>
+                <p>Be sure to select Settings to specify a Display Name and other values that apply.</p>
```
Bold Settings and Display Name (throughout) if possible

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/problem_with_hint.yaml:
```diff
@@ -4,13 +4,12 @@ metadata:
-    This problem demonstrates a question with hints, based on using the <tt class="tt">hintfn</tt> method. </p>
-        <script type="text/python" system_path="python_lib">
+        <text>
+            <p><h4>Problem With Adaptive Hint</h4></p>
```
I know you're bringing the same tagging forward, just changing the line feeds and indentation, but is it really necessary to have both the < p > and < h4 > tags for this line?

lamagnifica commented on common/lib/xmodule/xmodule/templates/problem/problem_with_hint_in_latex.yaml:
```diff
@@ -49,12 +49,11 @@ metadata:
-        <p>
-    This problem demonstrates a question with hints, based on using the <tt class="tt">hintfn</tt> method. </p>
+        <text>
+            <p><h4>Problem With Adaptive Hint</h4></p>
```
same question

lamagnifica commented:
Thank you for giving me the opportunity to review! I only looked at the yaml files, and commented on everything that I noticed.

lamagnifica commented:
@muhammad-ammar and @cahrens and all other interested parties, how about these alternatives (using quotation marks and a plain text reference to the documentation URL):

For the href:

You can refer learners to the "Entering Mathematical and Scientific Expressions" section in the "EdX Guide for Students" (https://docs.edx.org) for more information.

For the bolding:

When you add a problem, be sure to select "Settings" to specify the display name that students will see.

muhammad-ammar commented:
@muzaffaryousaf I am done with all the changes. I will do template content related changes in https://openedx.atlassian.net/browse/TNL-4977

muhammad-ammar commented:
jenkins run bokchoy
